src/opentime/rationalTime.cpp

```diff
@@ -205,55 +268,88 @@ RationalTime::from_timecode(
     return RationalTime{ double(value), rate };
 }
 
+static void
+set_error(std::string const& time_string,
+          ErrorStatus::Outcome code,
+          ErrorStatus* err)
+{
+    if (err) {
+        *err = ErrorStatus(
+            code,
+            string_printf(
+                "Input time string '%s' is an invalid time string",
+                time_string.c_str()));
+    }
+}
+
 RationalTime
 RationalTime::from_time_string(
     std::string const& time_string, double rate, ErrorStatus* error_status)
 {
-    if (!RationalTime::is_valid_timecode_rate(rate))
-    {
-        if (error_status)
-        {
-            *error_status = ErrorStatus(ErrorStatus::INVALID_TIMECODE_RATE);
-        }
+    if (!RationalTime::is_valid_timecode_rate(rate)) {
+        set_error(time_string, ErrorStatus::INVALID_TIMECODE_RATE, error_status);
         return RationalTime::_invalid_time;
     }
 
-    std::vector<std::string> fields(3, std::string());
-
-    // split the fields
-    int last_pos = 0;
-
-    for (int i = 0; i < 2; i++)
-    {
-        fields[i] = time_string.substr(last_pos, 2);
-        last_pos  = last_pos + 3;
-    }
-
-    fields[2] = time_string.substr(last_pos, time_string.length());
-
-    double hours, minutes, seconds;
+    const char* start = time_string.data();
+    const char* end = start + time_string.length();
+    char* current = const_cast<char*>(end);
+    char* parse_end = current;
+    char* prev_parse_end = current;
+
+    double power[3] = {
+        1.0,   // seconds
+        60.0,  // minutes
+        3600.0 // hours
+    };
+
+    double accumulator = 0.0;
+    int radix = 0;
+    while (start <= current) {
+        if (*current == ':') {
+            parse_end = current + 1;
+            char c = *parse_end;
+            if (c != '\0' && c != ':') {
+                if (c< '0' || c > '9') {
+                    set_error(time_string, ErrorStatus::INVALID_TIME_STRING, error_status);
+                    return RationalTime::_invalid_time;
+                }
+                double val = 0.0;
+                parseFloat(parse_end, prev_parse_end + 1, false, &val);
```

Comment:
The success/fail return value of parseFloat should be checked and turned into a set_error call, right? Same thing for the 2nd use of parseFloat below.

Reply:
done